httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/AIMDBackoffManager.java

```diff
@@ -56,79 +62,134 @@
+        this.lastRouteBackoffs = new ConcurrentHashMap<>();
     }
 
     @Override
     public void backOff(final HttpRoute route) {
-        synchronized(connPerRoute) {
-            final int curr = connPerRoute.getMaxPerRoute(route);
-            final Long lastUpdate = getLastUpdate(lastRouteBackoffs, route);
-            final long now = clock.getCurrentTime();
-            if (now - lastUpdate < coolDown.toMilliseconds()) {
-                return;
+        final int curr = connPerRoute.getMaxPerRoute(route);
+        final Instant now = Instant.now();
+
+        lastRouteBackoffs.compute(route, (r, lastUpdate) -> {
+            if (lastUpdate == null || now.isAfter(lastUpdate.plus(Duration.ofMillis(coolDown.get().toMilliseconds())))) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Backoff applied for route: {}, new max connections: {}", route, connPerRoute.getMaxPerRoute(route));
+                }
+                connPerRoute.setMaxPerRoute(route, getBackedOffPoolSize(curr));
+                return now;
             }
-            connPerRoute.setMaxPerRoute(route, getBackedOffPoolSize(curr));
-            lastRouteBackoffs.put(route, now);
-        }
+            return lastUpdate;
+        });
     }
 
+
+    /**
+     * Returns the backed-off pool size based on the current pool size.
+     * The new pool size is calculated as the floor of (backoffFactor * curr).
+     *
+     * @param curr the current pool size
+     * @return the backed-off pool size, with a minimum value of 1
+     */
     private int getBackedOffPoolSize(final int curr) {
         if (curr <= 1) {
             return 1;
         }
-        return (int)(Math.floor(backoffFactor * curr));
+        return (int) (Math.floor(backoffFactor.get() * curr));
     }
 
+
     @Override
     public void probe(final HttpRoute route) {
-        synchronized(connPerRoute) {
-            final int curr = connPerRoute.getMaxPerRoute(route);
-            final int max = (curr >= cap) ? cap : curr + 1;
-            final Long lastProbe = getLastUpdate(lastRouteProbes, route);
-            final Long lastBackoff = getLastUpdate(lastRouteBackoffs, route);
-            final long now = clock.getCurrentTime();
-            if (now - lastProbe < coolDown.toMilliseconds()
-                || now - lastBackoff < coolDown.toMilliseconds()) {
-                return;
+        final int curr = connPerRoute.getMaxPerRoute(route);
+        final int max = (curr >= cap.get()) ? cap.get() : curr + 1;
+        final Instant now = Instant.now();
+
+        lastRouteProbes.compute(route, (r, lastProbe) -> {
+            if (lastProbe == null || now.isAfter(lastProbe.plus(Duration.ofMillis(coolDown.get().toMilliseconds())))) {
+                final Instant lastBackoff = lastRouteBackoffs.get(r);
+                if (lastBackoff == null || now.isAfter(lastBackoff.plus(Duration.ofMillis(coolDown.get().toMilliseconds())))) {
```

Comment:
@arturobernalg It is a nitpick but one can avoid conversion to the `Duration` here. I think `lastProbe.plus(coolDown.get().toMilliseconds(), ChronoUnit.MILLIS)` should work the same way here with one less intermediate object on the heap.

Reply:
Sounds reasonable.
changed.